spark/src/main/java/org/apache/iceberg/spark/data/SparkValueReaders.java

```diff
@@ -86,8 +88,8 @@ static ValueReader<InternalRow> struct(List<ValueReader<?>> readers, Types.Struc
     return new StructReader(readers, struct, idToConstant);
   }
 
-  static ValueReader<InternalRow> union(Schema schema, List<ValueReader<?>> readers) {
-    return new UnionReader(schema, readers);
+  static ValueReader<InternalRow> union(Schema schema, List<ValueReader<?>> readers, Type expected) {
+    return new UnionReader(schema, readers, expected);
```

Comment:
Could you add a comment on why `expected` is needed in the union case, but not all the other readers? It is an interesting change since it added `Type` as a dependency of this class, which was not the case before.